File: Java/Visitor/ElectronicDevices/src/com/ericgtkb/ElectronicDevice.java
package com.ericgtkb;

public interface ElectronicDevice {
    void accept(DeviceVisitor visitor);
}

class Laptop implements ElectronicDevice {
    private double price;
    private double power;

    public Laptop() {
        price = 499.99;
        power = 70;
    }

    public double getPrice() {
        return price;
    }

    public double getPower() {
        return power;
    }

    @Override
    public void accept(DeviceVisitor visitor) {
        visitor.visit(this);
    }
}

class TV implements ElectronicDevice {
    private double price;
    private double power;

    public TV() {
        price = 339.99;
        power = 140;
    }

    public double getPrice() {
        return price;
    }

    public double getPower() {
        return power;
    }

    @Override
    public void accept(DeviceVisitor visitor) {
        visitor.visit(this);
    }
}

class VideoGameConsole implements ElectronicDevice {
    private double price;
    private double power;

    public VideoGameConsole() {
        price = 435.95;
        power = 12;
    }

    public double getPrice() {
        return price;
    }

    public double getPower() {
        return power;
    }

    @Override
    public void accept(DeviceVisitor visitor) {
        visitor.visit(this);
    }
}

File: C++/Visitor/ElectronicDevices/electronic_device.cc
#include "electronic_device.h"

Laptop::Laptop() : _price(499.99), _power(70) {}

double Laptop::get_price() const {
    return _price;
}

double Laptop::get_power() const {
    return _power;
}

void Laptop::accept(DeviceVisitor& visitor) {
    visitor.visit(this);
}


TV::TV() : _price(339.99), _power(140) {}

double TV::get_price() const {
    return _price;
}

double TV::get_power() const {
    return _power;
}

void TV::accept(DeviceVisitor& visitor) {
    visitor.visit(this);
}


VideoGameConsole::VideoGameConsole() : _price(435.95), _power(12) {}

double VideoGameConsole::get_price() const {
    return _price;
}

double VideoGameConsole::get_power() const {
    return _power;
}

void VideoGameConsole::accept(DeviceVisitor& visitor) {
    visitor.visit(this);
}
